### ISSUE TITLE
Proxy class can not be written to file, when it was already assembled before

### ISSUE THREAD

**Issue by holtkamp** [Ocramius/ProxyManager#364]
As described here https://github.com/PHP-DI/PHP-DI/issues/459, the configuration to write proxies to a file is not considered when generating the proxy class name (especially the hash of the proxy parameters). 

Not sure it *should* be, but this makes it impossible to write a proxy to a file, when it was already 'loaded' once with a configuration to not write it to a file. 

This typically happens when one wants to generate proxy files in a development environment / as part of a deploy process, where some proxies are already loaded during bootstrapping of the functionality that should actually generate the proxies... (hope that is not too abstract 😄 )

**Comment by Ocramius** [comment]
@holtkamp it is a bit abstract. In general, you want to consider the current `ProxyManager\Config` as part of the hash, if I understand it correctly.

That seems reasonable, since multiple running applications may use their proxies from different locations.

Unsure if worth fixing though, as nothing in a proxy class definition depends on the `ProxyManager\Config`. Two calls to a proxy factory, using the same settings, yet different `ProxyManager\Config`, should produce the exact same class.

If I understand it correctly, if you were to write a test for it, it would be as following (currently failing):

```php
$config1 = new \ProxyManager\Config();
$config2 = new \ProxyManager\Config();
// $config1->setProxiesToBeEvaluatedInMemory();
// $config2->setProxiesToBeWrittenToDisk();

$proxy1 = (new SomeFactory($config1))->createProxy(Foo::class);
self::assertFileNotExists((new ReflectionClass($proxy1))->getFileName());
$proxy2 = (new SomeFactory($config2))->createProxy(Foo::class);
self::assertFileExists((new ReflectionClass($proxy2))->getFileName());
```

Is that correct?

**Comment by holtkamp** [comment]
>Is that correct?
Exactly, writing files to disk is currently impossible when it was not the first time the proxy is generated. 

>Unsure if worth fixing though, as nothing in a proxy class definition depends on the ProxyManager\Config
I would say it is worth fixing (at least for the use-case indicated in https://github.com/PHP-DI/PHP-DI/issues/459), only the question is whether this is the proper location to fix it... I think it is the simplest way to let `class_exists()` return false...
